R/test-compiled-code.r

```diff
@@ -0,0 +1,224 @@
+  # variables in a configure script, but it's good for 99%
+  # of users. We list the files in descending order of
+  # priority.
+  makevars_paths <- c(
+
+    # Makevars.site in etc
+    file.path(R.home(), "etc", "Makevars.site"),
+
+    # ~/.R/Makevars (note: same on windows)
+    file.path(Sys.getenv("HOME"), ".R", "Makevars"),
+
+    # src/Makevars
+    if (is_windows)
+      file.path(pkg_path, "src", "Makevars")
+    else
+      file.path(pkg_path, "src", "Makevars.win")
+
+  )
+
+  flags <- lapply(makevars_paths, function(path) {
+    if (!file.exists(path)) return(character())
+
+    content <- paste(readLines(path), collapse = "\n")
+
+    # Replace '\\n' with plain old space
+    content <- gsub("\\\\\\s*\\n", " ", content, perl = TRUE)
+
+    # Now split back up again -- now, each unique definition
+    # should be on its own line
+    splat <- strsplit(content, "\n", fixed = TRUE)[[1]]
+
+    # Find the first '=' on each line
+    splat <- grep("=", splat, fixed = TRUE, value = TRUE)
+    equals_indices <- regexpr("=", splat, fixed = TRUE)
+
+    # Split into key, value pairs
+    keys <- substring(splat, 1, equals_indices - 1)
+    values <- substring(splat, equals_indices + 1, nchar(splat))
+
+    # Strip a trailing ':' off of the keys (in case of FOO := BAR)
+    keys <- sub(":$", "", keys, perl = TRUE)
+
+    # Trim off leading PKG_ qualifiers (we want to
+    # respect those)
+    keys <- gsub("^PKG_", "", keys, perl = TRUE)
+
+    # Normalize the keys and values
+    keys <- trim_whitespace(keys)
+    values <- trim_whitespace(values)
+
+    names(values) <- keys
+    values
+
+  })
+
+  # Collapse the set of variables
+  collapse <- function(x, y) {
+    for (i in seq_along(y))
+      x[[names(y)[i]]] <- y[[i]]
+    x
+  }
+
+  result <- as.list(Reduce(collapse, flags))
+
+  # If we have CXX_STD = CXX11, or "SystemRequirements: C++11",
+  # make sure we put -std=c++11 in the CXXFLAGS
+  DESCRIPTION <- tryCatch(
+    read.dcf(file.path(pkg_path, "DESCRIPTION"), all = TRUE),
+    error = function(e) NULL
+  )
+
+  SystemRequirements <- if (length(DESCRIPTION))
+    trim_whitespace(DESCRIPTION[["SystemRequirements"]])
+
+  if (identical(result[["CXX_STD"]], "CXX11") ||
+      identical(SystemRequirements, "C++11"))
+  {
+    result[["CXXFLAGS"]] <-
+      trim_whitespace(paste(result[["CXXFLAGS"]], "-std=c++11"))
+  }
+
+  result
+
+}
+
+parse_linking_to <- function(pkg_path) {
+
+  if (!file.exists(file.path(pkg_path, "DESCRIPTION")))
+    stop("No 'DESCRIPTION' file for directory '", basename(pkg_path), "'")
+
+  DESCRIPTION <- read.dcf(file.path(pkg_path, "DESCRIPTION"), all = TRUE)
+  LinkingTo <- DESCRIPTION$LinkingTo
+
+  if (is.null(LinkingTo))
+    return(character())
+
+  splat <- strsplit(LinkingTo, "\\s*[,\\n]\\s*", perl = TRUE)[[1]]
+
+  # Remove any version information
+  splat <- gsub("\\s*\\(.*", "", splat)
+
+  splat
+
+}
+
+test_compiled_code <- function(package, test_path, filter, verbose = FALSE) {
```

Comment:
I think you can replace 100% of the test_complied_code implementation by just calling R CMD SHLIB with the test_path as an argument, etc. (it will read Makevars, build the command line, etc.). You *might* still need to do the DESCRIPTION parsing but command line flags will all be gotten right via R CMD SHLIB. Note that this in the technique we use for libclang in RStudio to discover all of the command line flags to feed to CRAN.